### DISCUSSION TITLE
NUnit lifecycle policy

### DISCUSSION THREAD

**Discussion by OsirisTerje** [nunit/nunit#4424] [Policies]
Back in the days NUnit was very concerned about keeping backwards compatibility wrt to target frameworks.  Only recently have we started to trim down on that.  
The new Microsoft lifecycle policies with a much higher cadence also makes it harder to maintain backwards compatibility all the way.  
We now have Net Framework at 4.6.2 as a minimum and .net standard 2.0 as the minimum for net core.
We have recently discussed to include .net 6.0 as a target.  If we keep the netstandard 2.0 we will then have 3 different targets (up from 2) in the package, meaning we "up'it" with 50%.
We do test for a multitude of targets to ensure it works on those, but the more we keep in there, the longer the tests takes, and the more verbose output we get (as many of the non-run tests still output a lot of noise).  
Going for a newer version of .net means we can also take advantage of the improvements there.  
Having to support older version using conditional compiling directives are a pain. Some we must accept, but can we shorten down the list it would help.  

* So, how long after Microsoft have dropped their support should we continue to support it in newer versions?  
* Should we have different lifecycle support for STS and LTS frameworks ? 
* Should we stop supporting netcore3.1 ?

Also see some discussions earlier in #3070 and #4416. 

Comments:  @manfred-brands @jnm2 @stevenaw @mikkelbu @SeanKilleen @rprouse

**Comment by rprouse** [comment]
I feel that our updates should track the lifecycle of the frameworks we are targeting, https://learn.microsoft.com/en-us/lifecycle/products/microsoft-net-and-net-core. My experience is that companies and developers that aren't keeping their projects up to date aren't updating their NuGet packages either.

I've also had to install older unsupported frameworks to make contributions to NUnit. I can do that on my personal computer, but I prefer to not do it on my work computer.

**Reply by manfred-brands** [reply]
I think @rprouse hit the nail. Older software that gets a bigfix here or there isn't updating to newer major versions of packages due to possible incompatibilities.

For active developments, our IT department insists that we upgrade as per MS lifecycle due to the stopping of security updates for out of support frameworks.
Because of that we dropped netcoreap3.1 and moved these to net6.0 with no issues.

**Comment by mikkelbu** [comment]
Both at my old company and at my existing company we have tried to update nuget packages as often as possible, but in both places we have also been "stuck" on old frameworks due to a lot of code and tests (i.e. updating from .NET Framework => .NET Core and updating from .NET Core 2.1 to .NET 6.0), so I think that there is a larger group that updates nuget packages, but which also are struggling to keep up with the MS release (for various reasons).

For the questions:
> So, how long after Microsoft have dropped their support should we continue to support it in newer versions?

I think we should keep the support for some time e.g. 6 months.

> Should we have different lifecycle support for STS and LTS frameworks ?

I think we should focus on LTS and only do the minimal wrt to STS to make tests run

> Should we stop supporting netcore3.1 ?

I think so

**Comment by stevenaw** [comment]
I think @mikkelbu 's thoughts more or less mirror my own, which makes this an easy comment to write :)

We update major versions when we "must" at my work, which means no less frequent than jumping from LTS to LTS as support is dropped for earlier versions. Conversely, I tend to install new frameworks on personal devices near the end of the public previews when it hits RC.

I also like the idea of dropping support for an EOL runtime in the framework up to 6 months after Microsoft's EOL date elapses. Anecdotally this has seemed to align with when cloud providers drop their own support, at least if past CI failures are any indication

**Comment by SeanKilleen** [comment]
Thanks all for this; sorry for weighing in late.

I think dropping support for unsupported frameworks in our subsequent major releases is a fine thing to do.

* Ecosystems move forward because people are incentivized to move them forward.
* Rob's point is spot on about the affected audience of these changes.
* For those that _are_ stuck, I think this gives them one more thing to point to in order to show that their orgs should move forward and treat their software more like a living, evolving value stream. 
* Those who need the newest OSS also, in my opinion, are opting into the social contract that accepts that OSS needs to move forward as well, which means they have to do their part to keep their system's basic platform up to date if they want the latest features from their libraries.
